File: android/src/main/kotlin/ch/waio/pro_video_editor/src/features/render/helpers/ApplyComposition.kt
package ch.waio.pro_video_editor.src.features.render.helpers

import android.content.Context
import androidx.media3.common.Effect
import androidx.media3.common.audio.AudioProcessor
import androidx.media3.common.util.UnstableApi
import androidx.media3.transformer.Composition
import ch.waio.pro_video_editor.src.features.render.models.RenderConfig
import java.io.File

/**
 * Result of building a composition: the composition itself plus a list of
 * temporary files that were created during the build (e.g. pre-rendered
 * audio WAVs). The caller MUST delete these temporary files after the
 * Transformer export finishes (success or failure).
 */
data class CompositionResult(
    val composition: Composition,
    val temporaryFiles: List<File>
)

/**
 * Creates a Media3 Composition from render configuration.
 *
 * This is a simplified wrapper function that delegates the actual work
 * to CompositionBuilder. The builder pattern provides better separation
 * of concerns and cleaner code organization.
 *
 * @param context Android context
 * @param config The render configuration containing all composition parameters
 * @param videoEffects List of video effects to apply (from EffectsProcessor)
 * @param audioEffects List of audio effects to apply (from EffectsProcessor)
 * @return [CompositionResult] with the composition and any temporary files
 *   created during the build, or null if no video clips were provided.
 */
@UnstableApi
fun applyComposition(
    context: Context,
    config: RenderConfig,
    videoEffects: List<Effect>,
    audioEffects: List<AudioProcessor>
): CompositionResult? {
    // Layered (multi-track) path: stack several video sequences on one canvas.
    config.composition?.let { composition ->
        // Global color filters / blur live in [videoEffects]; image overlays are
        // converted here and applied at the composition level.
        val imageLayerConfigs = config.imageLayers.map { imageLayer ->
            VideoSequenceBuilder.ImageLayerConfig(
                imageBytes = imageLayer.imageData,
                scaleX = config.scaleX,
                scaleY = config.scaleY,
                withCropping = config.imageBytesWithCropping,
                startUs = imageLayer.startUs,
                endUs = imageLayer.endUs,
                x = imageLayer.x,
                y = imageLayer.y,
                width = imageLayer.width,
                height = imageLayer.height,
                rotation = imageLayer.rotation,
                animations = imageLayer.animations
            )
        }
        val layeredBuilder = LayeredCompositionBuilder(
            context = context,
            config = composition,
            enableAudio = config.enableAudio,
            globalVideoEffects = videoEffects,
            imageLayers = imageLayerConfigs,
            audioTracks = config.audioTracks,
            globalStartUs = config.startUs,
            globalEndUs = config.endUs
        )
        return CompositionResult(
            layeredBuilder.build(),
            layeredBuilder.temporaryFiles.toList()
        )
    }

    val builder = CompositionBuilder(context, config)
        .setVideoEffects(videoEffects)
        .setAudioEffects(audioEffects)
    val composition = builder.build() ?: return null
    return CompositionResult(composition, builder.temporaryFiles.toList())
}

File: darwin/pro_video_editor/Sources/pro_video_editor/src/shared/features/render/helpers/ApplyComposition.swift
import AVFoundation
import Foundation

/// Creates a multi-clip video composition with audio mixing and custom effects.
///
/// This is a simplified wrapper function that delegates the actual work
/// to CompositionBuilder. The builder pattern provides better separation
/// of concerns and cleaner code organization.
///
/// - Parameters:
///   - videoClips: Array of video clips to concatenate. Each clip can have optional trimming.
///   - videoEffects: Configuration for visual effects (rotation, scale, color, blur, etc.).
///   - enableAudio: If true, includes original audio from video clips.
///   - audioTracks: Array of audio track configurations to mix over the video.
///
/// - Returns: A tuple containing:
///   - AVMutableComposition: The concatenated video/audio composition
///   - VideoCompositionData: Video composition data with instructions and render size
///   - CGSize: Final render size (max dimensions from all clips)
///   - AVAudioMix?: Audio mix with volume controls (nil if no audio mixing needed)
///   - CMPersistentTrackID: The track ID of the video composition track (for fallback on older iOS)
///   - [URL]: Temporary file URLs (e.g. pre-rendered audio WAVs) the caller MUST delete after export
///   - [FadeWindow]: Dip-to-color windows for fadeToBlack / fadeToWhite transitions
///
/// - Throws: NSError if video clips are empty, files don't exist, or tracks can't be loaded.
func applyComposition(
  videoClips: [VideoClip],
  videoEffects: VideoCompositorConfig,
  enableAudio: Bool,
  audioTracks: [AudioTrackConfig]
) async throws -> (
  AVMutableComposition, VideoCompositionData, CGSize, AVAudioMix?, CMPersistentTrackID, [URL],
  [FadeWindow]
) {
  return try await CompositionBuilder(videoClips: videoClips, videoEffects: videoEffects)
    .setEnableAudio(enableAudio)
    .setAudioTracks(audioTracks)
    .build()
}
